### PR TITLE
HDDS-1535. Space tracking for Open Containers : Handle Node Startup. Contributed by Supratim Deka

### PR DESCRIPTION
**Pull request by supratimdeka**
https://issues.apache.org/jira/browse/HDDS-1535

Follow up from HDDS-1511.
Space tracking for Open Containers (committed space in the volume) relies on usedBytes in the Container state. usedBytes is not persisted for every update (chunkWrite). So on a node restart the value is stale.

The fix iterates the block DB for Open containers when building the container set (during startup).

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 27 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 397 | trunk passed |
| +1 | compile | 202 | trunk passed |
| +1 | checkstyle | 49 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 823 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 127 | trunk passed |
| 0 | spotbugs | 236 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 415 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 396 | the patch passed |
| +1 | compile | 205 | the patch passed |
| +1 | javac | 205 | the patch passed |
| -0 | checkstyle | 30 | hadoop-hdds: The patch generated 2 new + 0 unchanged - 0 fixed = 2 total (was 0) |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 630 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 112 | the patch passed |
| +1 | findbugs | 429 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 154 | hadoop-hdds in the patch failed. |
| -1 | unit | 1205 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 30 | The patch does not generate ASF License warnings. |
| | | 7313 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.client.rpc.TestBlockOutputStreamWithFailures |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-832/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/832 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux c94d1abf2ea8 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 732133c |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-832/1/artifact/out/diff-checkstyle-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-832/1/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-832/1/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-832/1/testReport/ |
| Max. process+thread count | 3953 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/container-service U: hadoop-hdds/container-service |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-832/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/ContainerReader.java [hunk 1]
```diff
@@ -215,4 +224,27 @@ public void verifyContainerData(ContainerData containerData)
...
+    KeyValueBlockIterator blockIter = new KeyValueBlockIterator(
+        container.getContainerData().getContainerID(),
+        new File(container.getContainerData().getContainerPath()));
+    long usedBytes = 0;
+
+    while (blockIter.hasNext()) {
+      BlockData block = blockIter.nextBlock();
+      long blockLen = 0;
+
+      List<ContainerProtos.ChunkInfo> chunkInfoList = block.getChunks();
```
Comment (arp7) [inline hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/ContainerReader.java]: Hi @supratimdeka , it looks like we are initializing using the metadata in RocksDB. I wonder if it is better to initialize this value using bytes on disk i.e. count usage of all the chunk files.

Both should match usually. However if there are chunk files that were written but never committed successfully then there could be a discrepancy. i.e. failed commit.

I don't think we are actively deleting such unreferenced chunk files currently, so if there are lot of failed writes the container could go far beyond its maximum capacity. This is a failure that could never be seen by DataNode so short of deleting the chunk files via scanner there is not much we can do.

Comment (arp7) [reply hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/ContainerReader.java]: Looked at this some more and even on the container write path the update is done on block commit. So this is consistent.

**Comment by arp7** [comment]
+1 pending CI checks.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 33 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 574 | trunk passed |
| +1 | compile | 264 | trunk passed |
| +1 | checkstyle | 70 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 863 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 147 | trunk passed |
| 0 | spotbugs | 313 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 524 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 533 | the patch passed |
| +1 | compile | 292 | the patch passed |
| +1 | javac | 292 | the patch passed |
| +1 | checkstyle | 73 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 651 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 138 | the patch passed |
| +1 | findbugs | 528 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 147 | hadoop-hdds in the patch failed. |
| -1 | unit | 1650 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 62 | The patch does not generate ASF License warnings. |
| | | 6716 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.client.rpc.TestOzoneClientRetriesOnException |
|   | hadoop.ozone.client.rpc.TestOzoneRpcClient |
|   | hadoop.ozone.TestMiniChaosOzoneCluster |
|   | hadoop.ozone.TestMiniOzoneCluster |
|   | hadoop.hdds.scm.pipeline.TestRatisPipelineProvider |
|   | hadoop.ozone.scm.TestAllocateContainer |
|   | hadoop.ozone.container.common.statemachine.commandhandler.TestBlockDeletion |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-832/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/832 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 9512ce78074a 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 9dff6ef |
| Default Java | 1.8.0_212 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-832/3/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-832/3/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-832/3/testReport/ |
| Max. process+thread count | 4196 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/container-service U: hadoop-hdds/container-service |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-832/3/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 29 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 536 | trunk passed |
| +1 | compile | 266 | trunk passed |
| +1 | checkstyle | 71 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 817 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 163 | trunk passed |
| 0 | spotbugs | 291 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 478 | trunk passed |
| -0 | patch | 359 | Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary. |
||| _ Patch Compile Tests _ |
| -1 | mvninstall | 105 | hadoop-hdds in the patch failed. |
| -1 | compile | 90 | hadoop-hdds in the patch failed. |
| -1 | javac | 90 | hadoop-hdds in the patch failed. |
| -0 | checkstyle | 45 | hadoop-hdds: The patch generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0) |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 681 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 162 | the patch passed |
| -1 | findbugs | 135 | hadoop-hdds in the patch failed. |
||| _ Other Tests _ |
| -1 | unit | 126 | hadoop-hdds in the patch failed. |
| -1 | unit | 1020 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 57 | The patch does not generate ASF License warnings. |
| | | 5868 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.container.common.impl.TestContainerPersistence |
|   | hadoop.ozone.client.rpc.TestBlockOutputStreamWithFailures |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-832/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/832 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 60e4fead4b43 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 03aa70f |
| Default Java | 1.8.0_212 |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-832/4/artifact/out/patch-mvninstall-hadoop-hdds.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-832/4/artifact/out/patch-compile-hadoop-hdds.txt |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-832/4/artifact/out/patch-compile-hadoop-hdds.txt |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-832/4/artifact/out/diff-checkstyle-hadoop-hdds.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-832/4/artifact/out/patch-findbugs-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-832/4/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-832/4/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-832/4/testReport/ |
| Max. process+thread count | 5059 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/container-service U: hadoop-hdds/container-service |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-832/4/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 50 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 1 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 545 | trunk passed |
| +1 | compile | 287 | trunk passed |
| +1 | checkstyle | 119 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 1032 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 177 | trunk passed |
| 0 | spotbugs | 448 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 652 | trunk passed |
| -0 | patch | 500 | Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary. |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 507 | the patch passed |
| +1 | compile | 303 | the patch passed |
| +1 | javac | 303 | the patch passed |
| +1 | checkstyle | 91 | the patch passed |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 763 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 164 | the patch passed |
| +1 | findbugs | 523 | the patch passed |
||| _ Other Tests _ |
| -1 | unit | 169 | hadoop-hdds in the patch failed. |
| -1 | unit | 1862 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 51 | The patch does not generate ASF License warnings. |
| | | 7456 | |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.ozone.client.rpc.TestBlockOutputStream |
|   | hadoop.ozone.container.common.impl.TestContainerPersistence |
|   | hadoop.ozone.scm.node.TestQueryNode |
|   | hadoop.ozone.client.rpc.TestBCSID |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-832/6/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/832 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 3c10e08de89b 4.4.0-144-generic #170~14.04.1-Ubuntu SMP Mon Mar 18 15:02:05 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / a771e2a |
| Default Java | 1.8.0_212 |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-832/6/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-832/6/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-832/6/testReport/ |
| Max. process+thread count | 5369 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/container-service U: hadoop-hdds/container-service |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-832/6/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by arp7** [comment]
The UT failures are unrelated. I've merged this into trunk.

**Pull request merged**